_posts/2020-11-23-how-fast-are-ractors.md

```diff
@@ -0,0 +1,133 @@
+
+All three benchmark types (fork, threads, Ractors) calculate the same things in the same way.
+
+The "fork" benchmark runs as a supervisor process, spawns N worker processes and talks to them over interprocess pipes. The "thread" benchmark uses a Mutex and a Ruby array as a queue. And the "ractor" benchmark uses [send/recv and yield/take](https://github.com/ruby/ruby/blob/master/doc/ractor.md) to pass results in the appropriate Ractor style.
+
+All results are sent back, sorted and checksummed to make sure all three benchmarks agree. The benchmark gets to time itself to see how long just the actual processing takes, and the [benchmark coordinator](https://github.com/noahgibbs/ractor_basic_benchmarks/tree/main/comparison_collector.rb) also times each process to see how long it takes with full setup and shutdown. The benchmarks are also run in a randomised order to keep varying load on the host-system load from turning into bias.
+
+And I originally timed this ***on my MacBook. Here's a tip: don't***. The performance and stability are substantially worse on Mac. I was getting around a 20% crash-and-error rate for medium-size benchmarks and some ***weird*** unexpected results. While things will probably be better by the release date, right now the Mac is ***not*** good for benchmarking Ractors.
+
+I'm still checking this on an itty bitty two-core instance, a Linode 4GB. This is a very prerelease feature. And based on the results I'm seeing, it's not ready for larger-scale benchmarking yet. That's okay, let's have a look anyway.
+
+There's some setup time for these benchmarks. If I run just 5 workers with 10 messages/worker plus startup and shutdown, here are the median times in seconds it takes for each benchmark: 
+
+### Initial Setup Overhead with Only 50 Total Reqs/Process
+
+<table>
+    <thead>
+        <tr>
+            <th></th><th>Messages-Only</th><th>Whole-Process</th>
+        </tr>
+    </thead>
+    <tbody>
+        <tr> <th>fork_test</th><td>0.012</td><td>0.123</td> </tr>
+        <tr> <th>thread_test</th><td>0.051</td><td>0.159</td> </tr>
+        <tr> <th>ractor_test</th><td>0.009</td><td>0.114</td> </tr>
+        <tr> <th>pipeless_ractor_test</th><td>0.009</td><td>0.109</td> </tr>
+    </tbody>
+</table>
+
+This is just the basic setup time. But if you wonder, "isn't there some fixed overhead before you're really timing anything?" &mdash; yes, yes there is. This table gives you an idea how much overhead.
+
+What's "pipeless?" In these tests, threads are coordinating directly with a queue. But there's an added bounce through the "pipe" for Ractors to allow them to finish at different speeds &mdash; that's how [Koichi's example code](https://github.com/ruby/ruby/blob/master/doc/ractor.md#worker-pool) did it. But I thought, maybe it's faster without it?
```

Comment:
should the `?` here be outside the quotes? `What's "pipeless"?`

Reply:
In programming, that's how it would work. In English-language style guides, the punctuation normally goes inside. I'm open to either one :-)

Reply:
My native language is Spanish so don't worry about that then haha. If that's the english style keep it like that.